torch/utils/data/dataset.py

```diff
@@ -91,8 +190,36 @@ def cummulative_sizes(self):
         return self.cumulative_sizes
 
 
-class Subset(Dataset):
+class ChainDataset(IterableDataset):
+    r"""Dataset for chainning multiple :class:`IterableDataset` s.
+
+    This class is useful to assemble different existing dataset streams. The
+    chainning operation is done on-the-fly, so concatenating large-scale
+    datasets with this class will be efficient.
+
+    Arguments:
+        datasets (iterable of IterableDataset): datasets to be chained together
     """
+    def __init__(self, datasets):
+        super(ChainDataset, self).__init__()
+        self.datasets = datasets
+
+    def __iter__(self):
+        for d in self.datasets:
+            assert isinstance(d, IterableDataset), "ChainDataset only supports IterableDataset"
+            for x in d:
+                yield x
```

Comment:
How would this chaining work? Don't we call iter() only once at the beginning of each worker loop?

Reply:
Could you elaborate why this would not work? This returns a generator, which is the iterable.

Reply:
I haven't used generators, so I might be missing something here. When you call `dataset_iter = iter(dataset)` in the worker_loop and the dataset is a ChainedDataset, isn't dataset_iter an iterator over a chain of IterableDataset and not really an iterator over the data itself?

Reply:
This goes through each `IterbleDataset` and yields each element from it. So it is yielding the data, not the dataset objects.

Reply:
This probably helps for understanding: https://docs.python.org/3/library/itertools.html#itertools.chain

Reply:
Oh my bad, for some reason I misread and thought it was yielding each dataset. Nevermind, this makes sense now.